### PR TITLE
Update `CONTRIBUTING.md` and support WSL2

### PR DESCRIPTION
**Pull request by compulim**
## Changelog Entry

(No log for doc and script change)

## Description

Updating our "too simple" `CONTRIBUTING.md`, and added support for WSL2-based Docker.

## Specific Changes

- Refreshed `docs/CONTRIBUTING.md`
- Fixed build break when developing under Linux
- Added `docker-compose.yml` for WSL2
- Added `package.json/scripts` for starting/stopping WSL2-based Docker images

---

-  [x] Testing Added
   - No tests added for doc/script change

### CODE REVIEW

File: .github/CONTRIBUTING.md [hunk 1]
```diff
@@ -8,64 +8,221 @@ When you submit a pull request, a CLA-bot will automatically determine whether y
 a CLA and decorate the PR appropriately (e.g., label, comment). Simply follow the instructions
 provided by the bot. You will only need to do this once across all repos using our CLA.
 
-# Building the project
+# Thank you
+
+Thanks for your interest in improving Web Chat. We invest heavily in engineering excellence to reduce our workloads and enable us to move faster. To start the development, please familiar yourself with the development process and these automation tools.
```
Comment (corinagum) [inline .github/CONTRIBUTING.md:13]:
```suggestion
Thanks for your interest in improving Web Chat. We invest heavily in engineering excellence to reduce our workloads and enable us to move faster. To start the development, please familiarize yourself with the development process and these automation tools.
```

File: .github/CONTRIBUTING.md [hunk 2]
```diff
@@ -8,64 +8,221 @@ When you submit a pull request, a CLA-bot will automatically determine whether y
...
+
+- Windows 10 Professional
+   - With WSL1 or [WSL2](https://aka.ms/wsl2) installed
+- [Ubuntu 18 running under WSL2](https://www.microsoft.com/en-us/p/ubuntu-1804-lts/9n9tngvndl3q)
 
-## Preparing the build
+## Preparing the repository
 
-After you clone the repository, run the following to make sure all dependencies were installed.
+Web Chat is a monorepo and use [`lerna`](https://lerna.js.org/). After the repository is [forked and cloned](https://github.com/microsoft/botframework-webchat/fork), run the following to install all dependencies.
```
Comment (corinagum) [inline .github/CONTRIBUTING.md:38]:
```suggestion
Web Chat is a monorepo and uses [`lerna`](https://lerna.js.org/). After the repository is [forked and cloned](https://github.com/microsoft/botframework-webchat/fork), run the following to install all dependencies.
```

File: .github/CONTRIBUTING.md [hunk 3]
```diff
@@ -8,64 +8,221 @@ When you submit a pull request, a CLA-bot will automatically determine whether y
...
 
-```sh
-cd packages/playground
-npm run start
-```
+- http://localhost:5000/webchat.js (also `webchat-es5.js` and `webchat-minimal.js`)
+- `packages/bundle/dist/webchat*.js`
 
-Then navigate to http://localhost:3000/, and click on one of the connection options on the upper right corner.
+# Playing the build
```
Comment (corinagum) [inline .github/CONTRIBUTING.md:59]:
```suggestion
# Playing with the build
```

File: .github/CONTRIBUTING.md [hunk 4]
```diff
@@ -8,64 +8,221 @@ When you submit a pull request, a CLA-bot will automatically determine whether y
...
 
--  Official MockBot: we hosted a live demo bot for testing Web Chat features
--  Emulator Core: it will connect to http://localhost:5000/v3/directline for [emulated Direct Line service](https://github.com/microsoft/BotFramework-Emulator/tree/master/packages/emulator/cli/)
+There are few ways you can play the build:
```
Comment (corinagum) [inline .github/CONTRIBUTING.md:61]:
```suggestion
There are few ways you can play with the build:
```

File: .github/CONTRIBUTING.md [hunk 5]
```diff
@@ -8,64 +8,221 @@ When you submit a pull request, a CLA-bot will automatically determine whether y
...
-npm run eslint
-```
+## Running automated tests
+
+Automated tests in Web Chat are designed to run inside a stable and isolated test environment. To run the test suite, you will need to [start the test environment](#starting-the-test-environment), then [run the test suite](#running-the-test-suite).
 
-eslint and prettier will ensure that your code follows our linting guidelines without too much effort. If you have any questions about this process, please feel free to leave comments in your Pull Request.
+### Starting the test environment
+
+For environment stability, Web Chat use Docker for hosting the test environment.
```
Comment (corinagum) [inline .github/CONTRIBUTING.md:89]:
```suggestion
For environment stability, Web Chat uses Docker for hosting the test environment.
```

File: .github/CONTRIBUTING.md [hunk 6]
```diff
@@ -8,64 +8,221 @@ When you submit a pull request, a CLA-bot will automatically determine whether y
...
 
-## Running integration tests
+# Readying for publication
 
-```bash
-docker-compose up --build --detach
-npm test
-docker-compose down --rmi all
-```
+We care about software quality. It helps prevent regressions, reduce maintenance costs, minimize chores, and enable us to move faster.
```
Comment (corinagum) [inline .github/CONTRIBUTING.md:76]:
```suggestion
We care about software quality. Quality checking prevents regressions, reduces maintenance costs, minimizes chores, and enables us to move faster.
```

File: .github/CONTRIBUTING.md [hunk 7]
```diff
@@ -8,64 +8,221 @@ When you submit a pull request, a CLA-bot will automatically determine whether y
...
 
-```
-npm run eslint
-```
+## Running automated tests
+
+Automated tests in Web Chat are designed to run inside a stable and isolated test environment. To run the test suite, you will need to [start the test environment](#starting-the-test-environment), then [run the test suite](#running-the-test-suite).
 
-eslint and prettier will ensure that your code follows our linting guidelines without too much effort. If you have any questions about this process, please feel free to leave comments in your Pull Request.
+### Starting the test environment
```
Comment (corinagum) [inline .github/CONTRIBUTING.md:87]: Make sure this info matches with `__tests__/README.md`. Might be good to consolidate.

File: .github/CONTRIBUTING.md [hunk 8]
```diff
@@ -8,64 +8,221 @@ When you submit a pull request, a CLA-bot will automatically determine whether y
...
+- Intermittent services issue
+- Test reliability issue, please see [#2938](https://github.com/microsoft/BotFramework-WebChat/issues/2938)
+- Polluted development environment, for example:
+   - Outdated `node_modules` content
+   - Outdated Node.js or NPM
+
+When the test suite failed:
+
+1. Re-run the test suite
+   1. If it succeeded but a specific test fail intermittently, please comment to [#2938](https://github.com/microsoft/BotFramework-WebChat/issues/2938)
```
Comment (corinagum) [inline .github/CONTRIBUTING.md:128]:
```suggestion
   1. If it succeeded but a specific test fails intermittently, please comment to [#2938](https://github.com/microsoft/BotFramework-WebChat/issues/2938)
```

File: .github/CONTRIBUTING.md [hunk 9]
```diff
@@ -8,64 +8,221 @@ When you submit a pull request, a CLA-bot will automatically determine whether y
...
+1. If it continue to fail, please fresh clone the repository and run the test suite without your changes
```
Comment (corinagum) [inline .github/CONTRIBUTING.md:129]:
```suggestion
1. If it continues to fail, please fresh clone the repository and run the test suite without your changes
```

File: .github/CONTRIBUTING.md [hunk 10]
```diff
@@ -8,64 +8,221 @@ When you submit a pull request, a CLA-bot will automatically determine whether y
...
+   1. If you suspect your environment is polluted, please use [Windows Sandbox](https://docs.microsoft.com/en-us/windows/security/threat-protection/windows-sandbox/windows-sandbox-overview) and/or a new Linux distro on WSL to verify
+1. If it still fail repeatedly, please [file an issue](https://github.com/microsoft/BotFramework-WebChat/issues/new/choose) to us.
```
Comment (corinagum) [inline .github/CONTRIBUTING.md:131]:
```suggestion
1. If it still fails repeatedly, please [file an issue](https://github.com/microsoft/BotFramework-WebChat/issues/new/choose) to us.
```

File: .github/CONTRIBUTING.md [hunk 11]
```diff
@@ -8,64 +8,221 @@ When you submit a pull request, a CLA-bot will automatically determine whether y
...
+
+## Static code analysis
 
-## Adding languages
+Run `npm run eslint` for static code analysis.
+
+If you need to skip any ESLint errors, we prefer using `eslint-disable-next-line`. Comments are required on why the rule is disabled.
```
Comment (corinagum) [inline .github/CONTRIBUTING.md:137]:
```suggestion
If you need to skip any ESLint errors, we recommend `eslint-disable-next-line`. Comments are required on why the rule is disabled.
```
Show example in the code?

File: .github/CONTRIBUTING.md [hunk 12]
```diff
@@ -8,64 +8,221 @@ When you submit a pull request, a CLA-bot will automatically determine whether y
...
+# Final checks
+
+There are checks that automation will not be able to capture. For example:
+
+- Hygiene
+   - Add a new log to `CHANGELOG.md`
+   - Fill out the pull request form for traceability
+   - Make sure imports, members, variables, etc, are sorted alphabetically
+   - Avoid one-off variables, prefer JavaScript shorthands, shorter and faster code
+   - No global pollution: no specific tab order, minimize `z-index` usage and requires a [new stacking context](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/The_stacking_context), all polyfills must complies to ECMAScript/W3C standards
```
Comment (corinagum) [inline .github/CONTRIBUTING.md:148]:
```suggestion
   - No global pollution: 
      - no specific tab order, 
      - minimize `z-index` usage
         - if z-index is ultimately required, use a [new stacking context](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/The_stacking_context), 
      - all polyfills must comply with ECMAScript/W3C standards

```

File: .github/CONTRIBUTING.md [hunk 13]
```diff
@@ -8,64 +8,221 @@ When you submit a pull request, a CLA-bot will automatically determine whether y
...
+- Tests
+   - Tests are important to reduce our burden. If new tests are needed but not included, we will reject your pull request
```
Comment (corinagum) [inline .github/CONTRIBUTING.md:150]:
```suggestion
   - Tests are important to reduces our maintenence burden. Pull requests without needed tests will not be approved or merged into the project.
```

File: .github/CONTRIBUTING.md [hunk 14]
```diff
@@ -8,64 +8,221 @@ When you submit a pull request, a CLA-bot will automatically determine whether y
...
+   - For fixing bug, the original bug repro must be coded as a new test
```
Comment (corinagum) [inline .github/CONTRIBUTING.md:151]:
```suggestion
   - For fixing a vbug, the original bug repro must be included as a new test
```

File: .github/CONTRIBUTING.md [hunk 15]
```diff
@@ -8,64 +8,221 @@ When you submit a pull request, a CLA-bot will automatically determine whether y
...
+   - For feature work, please add as much tests as needed to future-proof the feature, include both happy and unhappy paths
```
Comment (corinagum) [inline .github/CONTRIBUTING.md:152]:
```suggestion
   - For feature work, please add as many tests as needed to future-proof the feature, including both happy and unhappy paths
```

File: .github/CONTRIBUTING.md [hunk 16]
```diff
@@ -8,64 +8,221 @@ When you submit a pull request, a CLA-bot will automatically determine whether y
...
+   - We prefer integration tests over unit tests
+   - Avoid using timeout in tests, use [fake timer](https://www.npmjs.com/package/lolex) instead
```
Comment (corinagum) [inline .github/CONTRIBUTING.md:154]:
```suggestion
   - Avoid using timeout in tests. Instead, use [fake timer](https://www.npmjs.com/package/lolex) instead
```

File: .github/CONTRIBUTING.md [hunk 17]
```diff
@@ -8,64 +8,221 @@ When you submit a pull request, a CLA-bot will automatically determine whether y
...
+   - Windows 10
+      - Chrome
+      - Edge Chromium
+      - Edge UWP
+      - Firefox
+      - IE11 (except speech features)
+   - Safari on macOS
+   - Safari on iOS or iPadOS
+   - Chrome on Android
+- Accessibility, please refer to [`docs/ACCESSIBILITY.md`](https://github.com/microsoft/BotFramework-WebChat/blob/master/docs/ACCESSIBILITY.md)
```
Comment (corinagum) [inline .github/CONTRIBUTING.md:171]:
```suggestion
- For accessibility, please refer to [`docs/ACCESSIBILITY.md`](https://github.com/microsoft/BotFramework-WebChat/blob/master/docs/ACCESSIBILITY.md)
```

File: .github/CONTRIBUTING.md [hunk 18]
```diff
@@ -8,64 +8,221 @@ When you submit a pull request, a CLA-bot will automatically determine whether y
...
+   - Tab order, content readability, assistive technology-only text, color contrast, etc.
```
Comment (corinagum) [inline .github/CONTRIBUTING.md:172]:
```suggestion
   - Tab order, content readability, assistive technology-only text, color contrast, etc. must be maintained.
```

File: .github/CONTRIBUTING.md [hunk 19]
```diff
@@ -8,64 +8,221 @@ When you submit a pull request, a CLA-bot will automatically determine whether y
...
+   - Assistive technology and browser compatibility
+      - NVDA: Chrome and Firefox
```
Comment (corinagum) [inline .github/CONTRIBUTING.md:174]: I think we need to include JAWS, even though it's not included in our audit.

File: .github/CONTRIBUTING.md [hunk 20]
```diff
@@ -8,64 +8,221 @@ When you submit a pull request, a CLA-bot will automatically determine whether y
...
+      - Narrator: Edge family and IE11
+      - VoiceOver: Safari
+      - TalkBack: Chrome on Android
+- Internationalization, please refer to [`docs/LOCALIZATION.md`](https://github.com/microsoft/BotFramework-WebChat/tree/master/docs/LOCALIZATION.md)
```
Comment (corinagum) [inline .github/CONTRIBUTING.md:178]:
```suggestion
- For internationalization, please refer to [`docs/LOCALIZATION.md`](https://github.com/microsoft/BotFramework-WebChat/tree/master/docs/LOCALIZATION.md)
```

File: .github/CONTRIBUTING.md [hunk 21]
```diff
@@ -8,64 +8,221 @@ When you submit a pull request, a CLA-bot will automatically determine whether y
...
+- Feature documentation, samples, live demo, operational
```
Comment (corinagum) [inline .github/CONTRIBUTING.md:179]: Not sure what you mean by 'operational' -- not enough context. Could you elaborate?

File: .github/CONTRIBUTING.md [hunk 22]
```diff
@@ -8,64 +8,221 @@ When you submit a pull request, a CLA-bot will automatically determine whether y
...
+   - All samples must also comes with hosted live demo
```
Comment (corinagum) [inline .github/CONTRIBUTING.md:180]:
```suggestion
   - All samples must also come with a hosted live demo
```

File: .github/CONTRIBUTING.md [hunk 23]
```diff
@@ -8,64 +8,221 @@ When you submit a pull request, a CLA-bot will automatically determine whether y
...
+1. Run the test, `npm run start:docker`, then `npm test -- --testPathPattern html/this-is-the-bug.html`
+   - Make sure the test fail as described in the repro
+1. Fix the bug
+1. Re-run `npm run start:docker` to refresh the code inside Docker
+1. Re-run the test, it should succeed
+   - Verify screenshots if any
+1. Add a log entry to `CHANGELOG.md`
+1. File a pull request
+
+## Implementing new feature
```
Comment (corinagum) [inline .github/CONTRIBUTING.md:204]:
```suggestion
## Implementing new features
```

File: .github/CONTRIBUTING.md [hunk 24]
```diff
@@ -8,64 +8,221 @@ When you submit a pull request, a CLA-bot will automatically determine whether y
...
+
+Write the user story while implementing the feature.
+
+1. Clone/prepare the repository or nuke an existing repository
+   - To nuke, run `npm run tableflip`
+1. Run continuous build, run `npm start`
+1. Clone `__tests__/html/simple.html` to `my-feature.html`, use it as a playground
```
Comment (corinagum) [inline .github/CONTRIBUTING.md:211]:
```suggestion
1. Clone `__tests__/html/simple.html` to `my-feature.html` and use it as a playground
```

File: package.json [hunk 25]
```diff
@@ -29,6 +29,10 @@
     "posteslint": "npm run prettier-readmes",
     "prettier-readmes": "prettier --write **/**/*.md --tab-width 3 --single-quote true",
     "start": "concurrently --kill-others --raw \"serve\" \"serve -p 5001 -c serve-test.json\" \"lerna run --ignore playground --parallel --stream start\"",
+    "start:docker": "npm run start:docker:build && npm run start:docker:up",
```
Comment (corinagum) [inline package.json:32]: What's your reasoning for providing these?

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 26]
```diff
@@ -1,18 +1,39 @@
-Fixes #<!-- If this addresses a specific issue, please provide the issue number here -->
+<!-- Please provide the issue number here if any -->
+
+> Fixes #
 
 ## Changelog Entry
 
- <!-- Please paste your new entry from CHANGELOG.MD here -->
+<!-- Please paste your new entry from CHANGELOG.MD here. Not needed for works related to development purpose only. -->
```
Comment (corinagum) [inline .github/PULL_REQUEST_TEMPLATE.md:7]:
```suggestion
<!-- Please paste your new entry from CHANGELOG.MD here. Entry is not required for work only related to development purposes. -->
```

File: .github/CONTRIBUTING.md (hunk 21 above)
Comment (compulim) [reply .github/CONTRIBUTING.md:184]: Updated to "operation of demo bots".

File: package.json (hunk 25 above)
Comment (compulim) [reply package.json:32]: Running Docker under WSL2 is slightly different when building and scaling out test agents. Devs will need to run 2 different scripts: `docker-compose -f docker-compose-wsl2.yml build --parallel` followed by `docker-compose -f docker-compose-wsl2.yml up --scale chrome=4`.

This `npm run start:docker` will run both scripts.

File: .github/CONTRIBUTING.md (hunk 7 above)
Comment (compulim) [reply .github/CONTRIBUTING.md:87]: Consolidated.

File: .github/CONTRIBUTING.md (hunk 19 above)
Comment (compulim) [reply .github/CONTRIBUTING.md]: Added JAWS for Chrome and Firefox.

**Pull request merged**